pennylane/ops/op_math/dot.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2018-2022 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains the definition of the dot function, which computes the dot product between
+a vector and a list of operators.
+"""
+from typing import Sequence
+
+import pennylane as qml
+from pennylane.operation import Operator
+from pennylane.ops import Hamiltonian
+
+
+def dot(coeffs: Sequence[float], ops: Sequence[Operator], hamiltonian=True):
+    r"""Returns the dot product between the ``coeffs`` vector and the ``ops`` list of operators.
+
+    This function returns the following linear combination: :math:`\sum_{k=0}^{N-1} c_k O_k`, where
+    :math:`c_k` and :math:`O_k` are the elements inside the ``coeffs`` and ``ops`` arguments respectively.
+
+    Args:
+        coeffs (Sequence[float]): sequence containing the coefficients of the linear combination
+        ops (Sequence[Operator]): sequence containing the operators of the linear combination
+        hamiltonian (bool, optional): If True, a :class:`Hamiltonian` operator is used to represent
+            linear combination. If False, a :class:`Sum` operator is returned. Defaults to True.
+
+    Raises:
+        ValueError: if the number of coefficients and operators does not match
+
+    Returns:
+        .Hamiltonian | .Sum | .Operator: operator describing the linear combination
+
+    **Example**
+
+    >>> coeffs = np.array([1.1, 2.2])
+    >>> ops = [qml.PauliX(0), qml.PauliY(0)]
+    >>> qml.ops.dot(coeffs, ops)
+    (1.1) [X0]
+    + (2.2) [Y0]
+    >>> qml.ops.dot(coeffs, ops, hamiltonian=False)
+    (1.1*(PauliX(wires=[0]))) + (2.2*(PauliY(wires=[0])))
+    """
+    if hamiltonian:
+        return Hamiltonian(coeffs=coeffs, observables=ops)
+    if qml.math.shape(coeffs)[0] != len(ops):
+        raise ValueError("Number of coefficients and operators does not match.")
+    operands = [qml.s_prod(coeff, op) if coeff != 1 else op for coeff, op in zip(coeffs, ops)]
+    return qml.op_sum(*operands) if len(operands) > 1 else operands[0]
```

Comment:
@AlbertMitjans out of curiosity: is there a way to instantiate the new linear combination operator _without_ performing a slow Python for loop? E.g., the way we can with the old `Hamiltonian` class?

If not we should consider if this is possible to do, as we should aim to bypass a for loop internally for performance reasons if we know the final form of the operator we are constructing. @Jaybsoni might also have some thoughts here re: Pauli sentences

Reply:
Right now there isn't. We were discussing last quarter with @Jaybsoni and @albi3ro of possible ways to implement this.

I believe one way is to add a new class to handle linear combinations of operators.

Reply:
> I believe one way is to add a new class to handle linear combinations of operators.

:+1: Hopefully this should start to come together with the new Pauli representations. Once we have an efficient method for _representing_ linear combinations, instantiating large Hamiltonians directly will be a clear performance win for end users (@trbromley)

Reply:
FYI, I benchmarked the instantiation time of the `Hamiltonian` class vs the `Sum` class:
```python
import time

from numpy.random import randint, random

import pennylane as qml

N = 100000
num_wires = 10
coeffs = (random(size=N) * 2) - 1
pauli_ops = [qml.PauliX, qml.PauliY, qml.PauliZ]
ops = [pauli_ops[randint(0, 3)](randint(0, num_wires)) for _ in range(N)]

t0 = time.time()
h1 = qml.ops.dot(coeffs, ops)
t1 = time.time()
print(f"Hamiltonian initialization time: {t1 - t0}")

t0 = time.time()
h2 = qml.ops.dot(coeffs, ops, hamiltonian=False)
t1 = time.time()
print(f"Sum initialization time: {t1 - t0}")
```
```
Hamiltonian initialization time: 0.200333833694458
Sum initialization time: 3.7425148487091064
```
Looks like the `for` loop decreases performance quite drastically for hamiltonians with a large number of coefficients and/or wires.

Reply:
Nice! yes this is inline with what I expected, always good to avoid Python for loops wherever we can. Especially if we already have the data that describes the object!

Reply:
Added a `dot` version for pauli arithmetic, which is **faster** than instantiating a `Hamiltonian`! See PR description for more information.

Reply:
> Nice! yes this is inline with what I expected, always good to avoid Python for loops wherever we can. Especially if we already have the data that describes the object!

The `for` loop is **not** the bottleneck here. The bottleneck is the `__init__.py` of operators.

Reply:
Oh! good to know :)